### ISSUE TITLE
[fs2.io.tcp] `interruptWhen` doesn't seem to terminate fs2.io.tcp streams?

### ISSUE THREAD

**Issue by kczulko** [functional-streams-for-scala/fs2#1300]
Hi,

I'm recently working with `fs2.io.tcp` module. I have two streams running concurrently (tcp server and tcp client) and I want to terminate my application with `SignallingRef` when some specific message will be received (sth similar like #1299 example which I made yesterday). I found that my app wasn't terminated as expected (when used `interruptWhen` mechanics). Sockets are being closed and the whole stream doesn't process any more messages but unfortunately the application doesn't terminate (console process is still running). I was able to fix that with additional `Stream.bracket(F.point(asynchronousChanngelGroup))(acg => F.point(acg.shutdown))` expression. 

Question here is about who is responsible for closing that resource? Is it `fs2.io.tcp` or the user who had to supply ACG to tcp stream? Maybe it is good to consider removing `AsynchronousChannelGroup` from `server`'s implicit parameter list, require it explicitly and/or somehow show the user who is responsible for closing this resource? I've found that in `fs2.io.tcp` test you are creating shutdown hook for it: https://github.com/functional-streams-for-scala/fs2/blob/series/1.0/io/src/test/scala/fs2/io/tcp/SocketSpec.scala#L23.
Or maybe it is expected behavior but it is good to point it out somewhere within the docs?

I do not have right now any example app that I can publish here. I can make one in the couple of days if you wish.

Thanks in advance,
Karol

**Comment by kczulko** [comment]
Here is the sample app for this issue:

```scala
package example

import java.net.{InetAddress, InetSocketAddress}
import java.nio.channels.AsynchronousChannelGroup
import java.nio.channels.spi.AsynchronousChannelProvider
import java.util.concurrent.Executors

import scala.concurrent.duration._

import cats.effect.{ExitCode, IO, IOApp}
import cats.syntax.all._
import fs2._
import fs2.concurrent.SignallingRef
import fs2.io.tcp._

object Fs2_1300 extends IOApp {

  implicit val acg: AsynchronousChannelGroup =
    AsynchronousChannelProvider
      .provider()
      .openAsynchronousChannelGroup(8, Executors.defaultThreadFactory())

  def tcpStream: Stream[IO, Stream[IO, Unit]] = {
    val port = 6666
    val address = "0.0.0.0"
    val isa = new InetSocketAddress(InetAddress.getByName(address), port)

    for {
      // FIXME: Uncomment this line if you want to terminate your app
      // _          <- Stream.bracket(IO.delay(acg))(asyncCg => IO.delay(asyncCg.shutdown))
      resource <- server[IO](bind = isa)
      socket     <- Stream.resource(resource)
      bufferSize = 16348
    } yield socket.reads(bufferSize)
        .through(text.utf8Decode)
        .through(text.lines)
        .map(println)
  }

  def sleepAndRaiseSignal(signal: SignallingRef[IO,Boolean]): Stream[IO, Stream[IO, Unit]] = {
    for {
      _ <- Stream.sleep(10.seconds)
      _  = println("Interrupting streams...")
      _ <- Stream.eval(signal.set(true))
    } yield Stream.empty
  }

  def run(args: List[String]): IO[ExitCode] = {
    val program = for {
      signal <- Stream.eval(SignallingRef[IO,Boolean](false))
      _      <- (tcpStream concurrently sleepAndRaiseSignal(signal))
                    .interruptWhen(signal)
                    .parJoin(3)
    } yield ()

    program.compile.drain.as(ExitCode.Success)
  }
}
```

**Comment by ChristopherDavenport** [comment]
If you leave non-daemon threads operational then an application cannot shut down. You can lift the ACG used into a resource such that it follows expected shutdown behavior. I have implement this in [ember](https://github.com/ChristopherDavenport/ember/blob/master/modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServer.scala#L30)

**Comment by mpilquist** [comment]
Fixed by the new `SocketGroup` type in 1.1.0-M1
